### ISSUE TITLE
Regression with recursive types

### ISSUE THREAD

**Issue by vmeurisse** [microsoft/pyright#3727]
**Describe the bug**
Since version 1.1.257, the type of `B.children` is no longer correctly handled in the code bellow. It was working fine until version 1.1.256.

With 1.1.256, I get as expected:
> test.py:14:23 - error: Cannot access member "method_that_dont_exist" for type "C"

With 1.1.257, I get:
> test.py:12:22 - error: "C" is not iterable (reportGeneralTypeIssues)

- The `method_that_dont_exist` is no longer detected
- pyright doesn't see the C type as an iterable

**To Reproduce**
Just run `pyright  test.py` without any config file.

**Code**
```python
from dataclasses import dataclass
from typing import List

class A:
    test: "C"

@dataclass
class B:
    children: "C"

    def test(self):
        for child in self.children:
            print(child)
        self.children.method_that_dont_exist()

C = List[B]
```

**Additional context**
Removing the unused class A or the dataclass annotation on B make the issue disappear.

**Comment by erictraut** [comment]
Thanks for the bug report — and for narrowing the problem down to a particular build. That made it easier to localize the bug.

This will be fixed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.263, which I just published.
